### PR TITLE
IssuingDistributionPoint support (parse only)

### PR DESCRIPTION
**Pull request by reaperhulk**
h/t to Irina Renteria for the initial work here

refs #2546

### CODE REVIEW

File: docs/x509/reference.rst [hunk 1]
```diff
@@ -2319,6 +2319,76 @@ X.509 Extensions
 
         :type: int
 
+.. class:: IssuingDistributionPoint(only_contains_user_certs,\
+           only_contains_ca_certs, indirect_crl, only_contains_attribute_certs,\
+           only_some_reasons=None, full_name=None, relative_name=None)
```
Comment (alex) [inline docs/x509/reference.rst]: We don't generally use defaults for these, why here?

File: docs/x509/reference.rst [hunk 2]
```diff
@@ -2319,6 +2319,76 @@ X.509 Extensions
...
+        :type: bool
+
+        Set to ``True`` if the CRL this extension is embedded within only
+        contains information about attribute certificates.
+
+    .. attribute:: only_some_reasons
+
+        :type: frozenset of :class:`ReasonFlags` or None
+
+        The reasons for which the issuing distribution point is valid.
```
Comment (alex) [inline docs/x509/reference.rst]: "None indicates that it's valid for ..."

File: src/cryptography/hazmat/backends/openssl/decode_asn1.py [hunk 3]
```diff
@@ -464,6 +464,28 @@ def _decode_general_subtrees(backend, stack_subtrees):
     return subtrees
 
 
+def _decode_issuing_dist_point(backend, idp):
+    idp = backend._ffi.cast("ISSUING_DIST_POINT *", idp)
+    idp = backend._ffi.gc(idp, backend._lib.ISSUING_DIST_POINT_free)
+    full_name = None
+    relative_name = None
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/decode_asn1.py]: Assign these in the else please

File: src/cryptography/hazmat/backends/openssl/decode_asn1.py [hunk 4]
```diff
@@ -464,6 +464,28 @@ def _decode_general_subtrees(backend, stack_subtrees):
...
+    if idp.distpoint != backend._ffi.NULL:
+        full_name, relative_name = _decode_distpoint(backend, idp.distpoint)
+
+    only_user = idp.onlyuser == 255
+    only_ca = idp.onlyCA == 255
+    indirect_crl = idp.indirectCRL == 255
+    only_attr = idp.onlyattr == 255
+    only_some_reasons = None
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/decode_asn1.py]: same

File: src/cryptography/x509/__init__.py [hunk 5]
```diff
@@ -51,6 +51,7 @@
 OID_FRESHEST_CRL = ExtensionOID.FRESHEST_CRL
 OID_INHIBIT_ANY_POLICY = ExtensionOID.INHIBIT_ANY_POLICY
 OID_ISSUER_ALTERNATIVE_NAME = ExtensionOID.ISSUER_ALTERNATIVE_NAME
+OID_ISSUING_DISTRIBUTION_POINT = ExtensionOID.ISSUING_DISTRIBUTION_POINT
```
Comment (alex) [inline src/cryptography/x509/__init__.py]: Are we exposing these? I thought this was just for backwards compat, and new ones should be on `ExtensionOID` only?

File: src/cryptography/x509/extensions.py [hunk 6]
```diff
@@ -1446,6 +1446,138 @@ def __repr__(self):
...
+                "unspecified and remove_from_crl are not valid reasons in an "
+                "IssuingDistributionPoint"
+            )
+
+        bools = [
+            only_contains_user_certs, only_contains_ca_certs,
+            indirect_crl, only_contains_attribute_certs
+        ]
+
+        if not all([isinstance(x, bool) for x in bools]):
```
Comment (alex) [inline src/cryptography/x509/extensions.py]: Drop the list-comprehension. Also I think this is easier to read if you just write out the `and`s

File: tests/x509/test_x509_ext.py [hunk 7]
```diff
@@ -4440,6 +4440,210 @@ def test_inhibit_any_policy(self, backend):
...
+class TestIssuingDistributionPointExtension(object):
+    @pytest.mark.parametrize(
+        ("filename", "expected"),
+        [
+            [
+                "crl_idp_fullname_indirect_crl.pem",
+                x509.IssuingDistributionPoint(
+                    False, False, True, False, None, [
+                        x509.UniformResourceIdentifier(
+                            u"http://myhost.com/myca.crl")], None)
```
Comment (alex) [inline tests/x509/test_x509_ext.py]: Please wrap this different so we don't have this rightward drift.

File: tests/x509/test_x509_ext.py [hunk 8]
```diff
@@ -4440,6 +4440,210 @@ def test_inhibit_any_policy(self, backend):
...
+            [TypeError, ('notabool', False, False, False, None, None, None)],
+            [TypeError, (False, 'notabool', False, False, None, None, None)],
+            [TypeError, (False, False, 'notabool', False, None, None, None)],
+            [TypeError, (False, False, False, 'notabool', None, None, None)],
+            [ValueError, (True, True, False, False, None, None, None)],
+            [ValueError, (False, False, True, True, None, None, None)],
+            [ValueError, (False, False, False, False, None, None, None)],
+        ]
+    )
+    def test_invalid_init(self, error, args):
```
Comment (alex) [inline tests/x509/test_x509_ext.py]: Let's name the args, for easier reading

File: docs/x509/reference.rst [hunk 9]
```diff
@@ -2319,6 +2319,77 @@ X.509 Extensions
 
         :type: int
 
+.. class:: IssuingDistributionPoint(only_contains_user_certs,\
+           only_contains_ca_certs, indirect_crl, only_contains_attribute_certs,\
+           only_some_reasons, full_name, relative_name)
```
Comment (alex) [inline docs/x509/reference.rst]: These should be in a different order, either the one from the RFC, or all the only "only"s togther, or something else, but right now they look basically random.

File: docs/x509/reference.rst [hunk 10]
```diff
@@ -2319,6 +2319,77 @@ X.509 Extensions
...
+
+    .. versionadded:: 2.5
+
+    Issuing distribution point is a CRL extension that identifies the CRL
+    distribution point and scope for a particular CRL, and it indicates whether
```
Comment (alex) [inline docs/x509/reference.rst]: Break the sentence at ", and", it's a run on ATM.

File: src/cryptography/hazmat/backends/openssl/decode_asn1.py [hunk 11]
```diff
@@ -464,6 +464,30 @@ def _decode_general_subtrees(backend, stack_subtrees):
...
+def _decode_issuing_dist_point(backend, idp):
+    idp = backend._ffi.cast("ISSUING_DIST_POINT *", idp)
+    idp = backend._ffi.gc(idp, backend._lib.ISSUING_DIST_POINT_free)
+    if idp.distpoint != backend._ffi.NULL:
+        full_name, relative_name = _decode_distpoint(backend, idp.distpoint)
+    else:
+        full_name = None
+        relative_name = None
+
+    only_user = idp.onlyuser == 255
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/decode_asn1.py]: oh my god, is this seriously how this works?

File: src/cryptography/x509/extensions.py [hunk 12]
```diff
@@ -1446,6 +1446,145 @@ def __repr__(self):
...
+                "indirect_crl and only_contains_attribute_certs "
+                "must all be boolean."
+            )
+
+        bools = [
+            only_contains_user_certs, only_contains_ca_certs,
+            indirect_crl, only_contains_attribute_certs
+        ]
+
+        if len([x for x in bools if x is True]) > 1:
```
Comment (alex) [inline src/cryptography/x509/extensions.py]: I think this is actually more clearly written as `if only_contains_user_certs + only_contains_ca_certs + indirect_crl + only_contains_attribute_certs > 1`?

File: src/cryptography/x509/extensions.py [hunk 13]
```diff
@@ -1446,6 +1446,145 @@ def __repr__(self):
...
+            raise ValueError(
+                "Only one of the following can be set to True: "
+                "only_contains_user_certs, only_contains_ca_certs, "
+                "indirect_crl, only_contains_attribute_certs"
+            )
+
+        if (
+            not any([
```
Comment (alex) [inline src/cryptography/x509/extensions.py]: can this just be a chain of `or`?

File: tests/x509/test_x509_ext.py [hunk 14]
```diff
@@ -4440,6 +4440,229 @@ def test_inhibit_any_policy(self, backend):
...
+    @pytest.mark.parametrize(
+        ("filename", "expected"),
+        [
+            [
+                "crl_idp_fullname_indirect_crl.pem",
+                x509.IssuingDistributionPoint(
+                    False, False, True, False, None, [
+                        x509.UniformResourceIdentifier(
+                            u"http://myhost.com/myca.crl")
+                    ], None
```
Comment (alex) [inline tests/x509/test_x509_ext.py]: The indentation on all these is super weird

Comment (alex) [reply tests/x509/test_x509_ext.py]: And please use kwargs, they're impossible to review without refereing to the docs 100 times ATM.

File: tests/x509/test_x509_ext.py [hunk 15]
```diff
@@ -4440,6 +4440,229 @@ def test_inhibit_any_policy(self, backend):
...
+        assert idp == expected
+
+    @pytest.mark.parametrize(
+        (
+            "error", "only_contains_user_certs", "only_contains_ca_certs",
+            "indirect_crl", "only_contains_attribute_certs",
+            "only_some_reasons", "full_name", "relative_name"
+        ),
+        [
+            [
```
Comment (alex) [inline tests/x509/test_x509_ext.py]: These should be tuples, not lists

File: src/cryptography/x509/extensions.py (hunk 13 above)
Comment (reaperhulk) [reply src/cryptography/x509/extensions.py]: Yes, do you think that's more clear than `any`?

Comment (alex) [reply src/cryptography/x509/extensions.py]: I think so. Do you disagree?

Comment (reaperhulk) [reply src/cryptography/x509/extensions.py]: I find it easier to quickly see that something must be truthy when `any` is present, but like most stylistic things I don't hold a strong opinion here.

File: src/cryptography/x509/extensions.py (hunk 12 above)
Comment (reaperhulk) [reply src/cryptography/x509/extensions.py]: Isn't that abusing the fact that a bool is secretly an int? `True + True + False == 2` is kind of a Python WTF

Comment (alex) [reply src/cryptography/x509/extensions.py]: Yes, it is. If you don't like that, at least drop the `is True` and just make this `if x`. And rename `bools` to something semantically meaningful.

File: src/cryptography/x509/extensions.py [hunk 16]
```diff
@@ -1446,6 +1446,136 @@ def __repr__(self):
...
+            isinstance(indirect_crl, bool) and
+            isinstance(only_contains_attribute_certs, bool)
+        ):
+            raise TypeError(
+                "only_contains_user_certs, only_contains_ca_certs, "
+                "indirect_crl and only_contains_attribute_certs "
+                "must all be boolean."
+            )
+
+        boolean_arguments = [
```
Comment (alex) [inline src/cryptography/x509/extensions.py]: Even more semantic! "crl_constraints" or something

**Pull request merged**